src/jit/flowgraph.cpp

```diff
@@ -25718,3 +25720,340 @@ bool Compiler::fgUseThrowHelperBlocks()
+                {
+                    fgTailMergeThrowsFallThroughHelper(predBlock, nonCanonicalBlock, canonicalBlock, predEdge);
+                    updateCount++;
+                }
+                break;
+
+                case BBJ_ALWAYS:
+                {
+                    fgTailMergeThrowsJumpToHelper(predBlock, nonCanonicalBlock, canonicalBlock, predEdge);
+                    updateCount++;
+                }
+                break;
+
+                case BBJ_COND:
+                {
+                    // Flow to non canonical block could be via fall through or jump or both.
+                    if (predBlock->bbNext == nonCanonicalBlock)
+                    {
+                        fgTailMergeThrowsFallThroughHelper(predBlock, nonCanonicalBlock, canonicalBlock, predEdge);
+                    }
+
+                    if (predBlock->bbJumpDest == nonCanonicalBlock)
+                    {
+                        fgTailMergeThrowsJumpToHelper(predBlock, nonCanonicalBlock, canonicalBlock, predEdge);
+                    }
+                    updateCount++;
+                }
+                break;
+
+                case BBJ_SWITCH:
+                {
+                    JITDUMP("*** " FMT_BB " now branching to " FMT_BB "\n", predBlock->bbNum, canonicalBlock->bbNum);
+                    fgReplaceSwitchJumpTarget(predBlock, canonicalBlock, nonCanonicalBlock);
+                    updateCount++;
+                }
+                break;
+
+                default:
+                    // We don't expect other kinds of preds, and it is safe to ignore them
+                    // as flow is still correct, just not as optimized as it could be.
+                    break;
+            }
+        }
+    }
+
+    // If we altered flow, reset fgModified. Given where we sit in the
+    // phase list, flow-dependent side data hasn't been built yet, so
+    // nothing needs invalidation.
+    //
+    // Note we could invoke a cleanup pass here, but optOptimizeFlow
+    // seems to be missing some safety checks and doesn't expect to
+    // see an already cleaned-up flow graph.
+    if (updateCount > 0)
+    {
+        assert(fgModified);
+        fgModified = false;
+    }
+
+#if DEBUG
+    if (verbose)
+    {
+        printf("\n*************** After fgTailMergeThrows(%d updates)\n", updateCount);
+        fgDispBasicBlocks();
+        fgVerifyHandlerTab();
+        fgDebugCheckBBlist();
+    }
+#endif // DEBUG
+}
+
+//------------------------------------------------------------------------
+// fgTailMergeThrowsFallThroughHelper: fixup flow for fall throughs to mergable throws
+//
+// Arguments:
+//    predBlock - block falling through to the throw helper
+//    nonCanonicalBlock - original fall through target
+//    canonicalBlock - new (jump) target
+//    predEdge - original flow edge
+//
+// Notes:
+//    Alters fall through flow of predBlock so it jumps to the canonicalBlock
+//    via a new basic block.
```

Comment:
Ideally there'd be a utility to call that would check for special restrictions, reverse the branch condition and make suitable flow updates. The logic for this is is in `fgUpdateFlowGraph` and nontrivial so I didn't want to duplicate code, and the blocks get updated (if suitable) right after this phase, during `optOptimizeFlow`.

I could refactor, if you think it makes sense to try and optimize this on the spot instead of waiting...

Reply:
No, it sounds like this is the right thing to do, but A comment to that effect would help explain why it's being done this way.